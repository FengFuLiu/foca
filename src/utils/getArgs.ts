const slice = Array.prototype.slice;

export const getArgs = <T = any>(args: IArguments): T[] => slice.call(args);
