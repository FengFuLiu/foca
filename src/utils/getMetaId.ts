export const getMetaId = (id: number | string) => '_' + id;
