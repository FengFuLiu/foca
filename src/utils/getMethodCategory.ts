export const getMethodCategory = (methodName: string) =>
  methodName.indexOf('_') === 0 ? 'hidden' : 'visible';
