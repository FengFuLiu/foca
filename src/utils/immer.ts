import { Immer, enableES5 } from 'immer';

/**
 * 支持ES5
 * @link https://immerjs.github.io/immer/docs/installation#pick-your-immer-version
 * @since immer 6.0
 */
enableES5();

export const immer = new Immer({
  autoFreeze: false,
});
