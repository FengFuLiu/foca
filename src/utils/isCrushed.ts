/*
 * The browsers which don't support Function.name should be considered as compressed.
 *
 * @link https://github.com/terser/terser
 * toplevel (default false) - set to true if you wish to enable top level variable
 * and function name mangling and to drop unused variables and functions.
 *
 * For ESM, don't put CustomFunction to toplevel.
 */
const name = (() => {
  function CustomFunction() {}
  return CustomFunction.name;
})();

const crushed = typeof name !== 'string' || name !== 'CustomFunction';

export const isCrushed = () => crushed;
