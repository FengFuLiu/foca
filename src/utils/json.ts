export const jsonStringifyReplacer = (_: string, value: any) => {
  switch (Object.prototype.toString.call(value).slice(8, -1)) {
    case 'Map':
      return `new Map(${JSON.stringify(
        Array.from((value as Map<any, any>).entries()),
      )})`;
    case 'Set':
      return `new Set(${JSON.stringify(
        Array.from((value as Set<any>).values()),
      )})`;
    default:
      return value;
  }
};

const toString = Object.prototype.toString;
const arrayString = '[object Array]';

export const jsonParseReciever = (_: string, value: any) => {
  if (typeof value === 'string') {
    switch (value.slice(0, 8)) {
      case 'new Map(':
        try {
          const maps = JSON.parse(value.slice(8, -1)) as [any, any][];

          if (
            toString.call(maps) === arrayString &&
            maps.every(
              (item) =>
                toString.call(item) === arrayString && item.length === 2,
            )
          ) {
            return new Map(maps);
          }
        } catch {}
        break;
      case 'new Set(':
        try {
          const sets = JSON.parse(value.slice(8, -1)) as any[];

          if (toString.call(sets) === arrayString) {
            return new Set(sets);
          }
        } catch {}
        break;
    }
  }

  return value;
};
