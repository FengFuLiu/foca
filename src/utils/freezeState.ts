import { freeze } from 'immer';

export const freezeState = <T = any>(state: T): T => {
  return process.env.NODE_ENV === 'production' ? state : freeze(state, true);
};
