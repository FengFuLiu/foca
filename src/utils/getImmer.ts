import { Immer, enableES5 } from 'immer';

const immer = new Immer({
  autoFreeze: false,
});

/**
 * support for the fallback implementation has to be explicitly enabled
 * @link https://immerjs.github.io/immer/docs/installation#pick-your-immer-version
 * @since immer 6.0
 */
enableES5();

export const getImmer = () => immer;
