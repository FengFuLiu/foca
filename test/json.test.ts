import isEqual from 'lodash.isequal';
import { jsonParseReciever, jsonStringifyReplacer } from '../src/utils/json';

test('json string and Map/Set can transform to each', () => {
  const map = new Map();
  map.set('aa', 'bb');
  map.set('cc', 2);

  const set = new Set();
  set.add(1);
  set.add(4);
  set.add('555');

  const obj = {
    a: 1,
    b: 'test-me',
    c: {
      def: 'ts2',
      test2: map,
      test5: {
        data: set,
      },
    },
    test3: set,
    test4: map,
  };

  const str = JSON.stringify(obj, jsonStringifyReplacer);

  expect(typeof str).toBe('string');
  expect(str).toMatchSnapshot('Map/Set string');

  const nextObj = JSON.parse(str, jsonParseReciever);
  expect(typeof nextObj).toBe('object');
  expect(nextObj).toMatchSnapshot('Map/Set object');
  expect(isEqual(obj, nextObj)).toBeTruthy();
});
