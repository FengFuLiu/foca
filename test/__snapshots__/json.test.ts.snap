// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`json string and Map/Set can transform to each: Map/Set object 1`] = `
Object {
  "a": 1,
  "b": "test-me",
  "c": Object {
    "def": "ts2",
    "test2": Map {
      "aa" => "bb",
      "cc" => 2,
    },
    "test5": Object {
      "data": Set {
        1,
        4,
        "555",
      },
    },
  },
  "test3": Set {
    1,
    4,
    "555",
  },
  "test4": Map {
    "aa" => "bb",
    "cc" => 2,
  },
}
`;

exports[`json string and Map/Set can transform to each: Map/Set string 1`] = `"{\\"a\\":1,\\"b\\":\\"test-me\\",\\"c\\":{\\"def\\":\\"ts2\\",\\"test2\\":\\"new Map([[\\\\\\"aa\\\\\\",\\\\\\"bb\\\\\\"],[\\\\\\"cc\\\\\\",2]])\\",\\"test5\\":{\\"data\\":\\"new Set([1,4,\\\\\\"555\\\\\\"])\\"}},\\"test3\\":\\"new Set([1,4,\\\\\\"555\\\\\\"])\\",\\"test4\\":\\"new Map([[\\\\\\"aa\\\\\\",\\\\\\"bb\\\\\\"],[\\\\\\"cc\\\\\\",2]])\\"}"`;
